src/sync.rs

```diff
@@ -0,0 +1,302 @@
+use std::collections::{hash_map::Entry, HashMap, VecDeque};
+use std::convert::TryFrom;
+use std::num::NonZeroU32;
+use std::time::Instant;
+
+use rustc_index::vec::{Idx, IndexVec};
+
+use crate::*;
+
+macro_rules! declare_id {
+    ($name: ident) => {
+        /// 0 is used to indicate that the id was not yet assigned and,
+        /// therefore, is not a valid identifier.
+        #[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+        pub struct $name(NonZeroU32);
+
+        impl $name {
+            // Panics if `id == 0`.
+            pub fn from_u32(id: u32) -> Self {
+                Self(NonZeroU32::new(id).unwrap())
+            }
+        }
+
+        impl Idx for $name {
+            fn new(idx: usize) -> Self {
+                $name(NonZeroU32::new(u32::try_from(idx).unwrap() + 1).unwrap())
+            }
+            fn index(self) -> usize {
+                usize::try_from(self.0.get() - 1).unwrap()
+            }
+        }
+
+        impl $name {
+            pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+                Scalar::from_u32(self.0.get())
+            }
+        }
+    };
+}
+
+declare_id!(MutexId);
+
+/// The mutex state.
+#[derive(Default, Debug)]
+struct Mutex {
+    /// The thread that currently owns the lock.
+    owner: Option<ThreadId>,
+    /// How many times the mutex was locked by the owner.
+    lock_count: usize,
+    /// The queue of threads waiting for this mutex.
+    queue: VecDeque<ThreadId>,
+}
+
+declare_id!(RwLockId);
+
+/// The read-write lock state.
+#[derive(Default, Debug)]
+struct RwLock {
+    /// The writer thread that currently owns the lock.
+    writer: Option<ThreadId>,
+    /// The readers that currently own the lock and how many times they acquired
+    /// the lock.
+    readers: HashMap<ThreadId, usize>,
+    /// The queue of writer threads waiting for this lock.
+    writer_queue: VecDeque<ThreadId>,
+    /// The queue of reader threads waiting for this lock.
+    reader_queue: VecDeque<ThreadId>,
+}
+
+declare_id!(CondvarId);
+
+/// A thread waiting on a conditional variable.
+#[derive(Debug)]
+struct CondvarWaiter {
+    /// The thread that is waiting on this variable.
+    thread: ThreadId,
+    /// The mutex on which the thread is waiting.
+    mutex: MutexId,
+    /// The moment in time when the waiter should time out.
+    timeout: Option<Instant>,
```

Comment:
Doesn't this also have to remember the clock used for that? If the system clock is used and it is changed after the mutex is created, I suppose that should be taken into account. (Not sure what should happen though when that is used to jump over the target time...)